arbnode/espresso_delayed_message_fetcher.go

```diff
@@ -143,97 +136,105 @@ func (d *DelayedMessageFetcher) processNewHeader(ctx context.Context, header *ty
 
-	// Get the delayed message count store in the database
-	delayedCount, err := getDelayedMessageLatestIndex(f.db)
-	if err != nil {
-		log.Error("Failed to get delayed message count from db", "err", err)
-		return nil, err
-	}
-
+	delayedCount := d.delayedCount
 	// If this is delayed message, we need to get the message from L1
 	// and replace the message in the messageWithMetadataAndPos
 	delayedMessageToProcess := delayedMessagesRead - 1
 
 	if delayedMessageToProcess > delayedCount {
 		log.Warn("delayed message fetcher is lagging behind", "delayedMessagesRead", delayedMessagesRead, "delayedCount", delayedCount)
-		return nil, fmt.Errorf("delayed message fetcher is lagging behind")
+		return nil, 0, fmt.Errorf("delayed message fetcher is lagging behind")
 	}
 	log.Debug("Getting delayed message", "delayedCount", delayedMessageToProcess)
 
 	// Note: here we are using DelayedMessagesRead - 1 because that is the index of the delayed message
 	// that needs to be read
-	message, err := f.readDelayedMessage(delayedMessageToProcess)
+	message, fromBlock, err := d.readDelayedMessageAndFromBlock(delayedMessageToProcess)
 	if err != nil {
 		log.Error("failed to get delayed message", "err", err)
-		return messageWithMetadataAndPos, err
+		return messageWithMetadataAndPos, 0, err
 	}
 	messageWithMetadataAndPos.MessageWithMeta.Message = message.Message
 
-	return messageWithMetadataAndPos, nil
+	return messageWithMetadataAndPos, fromBlock, nil
 }
 
-/***** Getter Functions *****/
-
 /*
 Reads the "current from" block from the database.
 */
-func readCurrentFromBlockFromDb(db ethdb.Database) (uint64, error) {
+func readCurrentFromBlockFromDb(db ethdb.Database) (uint64, []byte, error) {
 	var blockNumber uint64
 	blockNumberBytes, err := db.Get([]byte(DelayedFetcherCurrentFromBlockKey))
 	if err != nil && !dbutil.IsErrNotFound(err) {
-		return 0, fmt.Errorf("failed to get next hotshot block: %w", err)
+		return 0, nil, fmt.Errorf("failed to get next from block: %w", err)
+	}
+	if dbutil.IsErrNotFound(err) {
+		return 0, nil, nil
 	}
 	if blockNumberBytes != nil {
 		err = rlp.DecodeBytes(blockNumberBytes, &blockNumber)
 		if err != nil {
-			return 0, fmt.Errorf("failed to decode next hotshot block: %w", err)
+			return 0, nil, fmt.Errorf("failed to decode next from block: %w", err)
 		}
 	}
 
-	return blockNumber, nil
+	// Also get the signature and check that signaure is valid
+	fromBlockSignatureBytes, err := db.Get(DelayedFetcherCurrentFromBlockSignatureKey)
+	if err != nil && !dbutil.IsErrNotFound(err) {
+		return 0, nil, fmt.Errorf("failed to get next from block signature: %w", err)
+	}
+	if dbutil.IsErrNotFound(err) {
+		return 0, nil, nil
+	}
+
+	log.Info("Read from block signature", "fromBlock", blockNumber, "fromBlockSignature", fromBlockSignatureBytes)
+	return blockNumber, fromBlockSignatureBytes, nil
+}
+
+func storeFromBlockWithSignature(batch ethdb.Batch, fromBlock uint64, fromBlockSignature []byte) error {
```

Comment:
Since this function, and similarly, `readCurrentBlockFromDb` are not functions specific to the `delayedMessageFetcher`, perhaps we should move them to espresso_utils.go?